### ISSUE TITLE
Replacing cling with clang-repl

### ISSUE THREAD

**Issue by a-turko** [wlav/cppyy#174]
Hi,
what do you think about replacing cling with [clang-repl](https://clang.llvm.org/docs/ClangRepl.html)?
From what I read ([Cling transitions to LLVM's clang-repl](https://clang.llvm.org/docs/ClangRepl.html)) ) it seems that clang-repl could be a good substitute.
This way cppyy would work with the newest clang.

Is something missing from clang-repl that prohibits the change?
If not, what would be your estimate of the effort needed for the transition?

**Comment by wlav** [comment]
Same folks behind both clang-repl and Cling. They're also working on this change, using the more generic libinterop, with currently 207 out of 504 cppyy tests passing. See [here](https://compiler-research.org/) for more information.

**Comment by a-turko** [comment]
Do you have an open task on that front that would take between 1 and 4 weeks for someone unfamiliar with the codebase?

**Comment by sudo-panda** [comment]
Hi @a-turko, thanks for your interest in the project.

You can have a look at these issues: 
 - https://github.com/compiler-research/cppyy-backend/issues
 - https://github.com/compiler-research/CPyCppyy/issues
 - https://github.com/compiler-research/CppInterOp/issues

The good first issues are marked there. For most of those issues, familiarity with the codebase is not required. Feel free to ask us if you need help setting up the dev environment. I suggest checking out the [CI on cppyy](https://github.com/compiler-research/cppyy/blob/master/.github/workflows/ci.yml) for up-to-date steps to set it up.

**Comment by vgvassilev** [comment]
> I was following the [CI config for cppyy](https://github.com/compiler-research/cppyy/blob/master/.github/workflows/ci.yml) for clang-repl and encountered an error when building CppInterOp: `CppInterOp/include/clang/Interpreter/Compatibility.h:56:10: fatal error: clang/Interpreter/Value.h: No such file or directory 56 | #include "clang/Interpreter/Value.h"`
> 
> As far as I can tell this file was never present in CppInterOp, but has been added (under the same path) to llvm, but only in newer releases. However, CI uses llvm16, where this files does not exist. I have consulted the instructions in the CppInterOp repository, but the result was the same.
> 
> Have you also run into similar issues? Could you please point me to a possible place where I could have misconfigured something?

You are right. That file will appear in llvm17. For now, we have patched it on top of llvm16 by backporting the patch. You can see in the CI how we apply these patches.

**Comment by a-turko** [comment]
Thank you. I see the `compgen -G "../patches/llvm/clang16-*.patch" > /dev/null && find ../patches/llvm/clang16-*.patch -printf "%f\n" && git apply ../patches/llvm/clang16-*.patch` command, but could you tell me where do you get the patches (../patches/llvm directory) from?

**Comment by sudo-panda** [comment]
> Thank you. I see the `compgen -G "../patches/llvm/clang16-*.patch" > /dev/null && find ../patches/llvm/clang16-*.patch -printf "%f\n" && git apply ../patches/llvm/clang16-*.patch` command, but could you tell me where do you get the patches (../patches/llvm directory) from?

You can find them in the CppInterOp repo: https://github.com/compiler-research/CppInterOp/tree/main/patches/llvm

**Comment by a-turko** [comment]
Thank you, now the installation went smoothly to the end. However, the command `python -c "import cppy"` failed with the following:

``
In file included from <<< inputs >>>:1:
input_line_2:1:10: fatal error: 'iostream' file not found
#include <iostream>
         ^~~~~~~~~~
python: /home/aturko/src/cling-front/llvm-project/clang/include/clang/AST/DeclBase.h:1302: clang::NamedDecl* clang::DeclListNode::iterator::operator*() const: Assertion `Ptr && "dereferencing end() iterator"' failed.
Aborted (core dumped)
``

Have you also run into this problem?

**Comment by vgvassilev** [comment]
We somehow did not find the path to the stl headers. Could you take a look at how the CI does it?

**Comment by a-turko** [comment]
I got it when following CI instructions. I'm not so sure about the directory structure though: should all dependencies be cloned directly into the cppyy repository (cppyy directory), or should they just have a commom parent, i.e. all repositories in the same directory?

**Comment by vgvassilev** [comment]
@a-turko, I believe if you cut and paste carefully these https://github.com/compiler-research/cppyy-backend/blob/4e244e486fb761dfa5559c95929d2d3b23535d0f/.github/workflows/ci.yml#L109-L258 it should work.

Generally, the error tells you that for some reason it cannot find `<iostream>`. Can you paste the output of `echo "" | c++ -fsyntax-only -xc++ - -v`?

**Comment by wlav** [comment]
Closing as stale, since no response to last request for info in over a year.
